WordPress/src/test/java/org/wordpress/android/viewmodel/themes/ThemesViewModelTest.kt

```diff
@@ -0,0 +1,144 @@
+package org.wordpress.android.viewmodel.themes
+
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import kotlinx.coroutines.InternalCoroutinesApi
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mock
+import org.wordpress.android.BaseUnitTest
+import org.wordpress.android.TEST_DISPATCHER
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.model.ThemeModel
+import org.wordpress.android.fluxc.store.ThemeStore
+import org.wordpress.android.viewmodel.ResourceProvider
+import org.wordpress.android.viewmodel.themes.ThemesViewModel.BottomSheetAction
+import org.wordpress.android.viewmodel.themes.ThemesViewModel.BottomSheetAction.Hide
+import org.wordpress.android.viewmodel.themes.ThemesViewModel.BottomSheetAction.Show
+import org.wordpress.android.viewmodel.themes.ThemesViewModel.BottomSheetUIState
+import org.wordpress.android.viewmodel.themes.ThemesViewModel.BottomSheetUIState.Selection.KeepCurrentHomepage
+import org.wordpress.android.viewmodel.themes.ThemesViewModel.BottomSheetUIState.Selection.UseThemeHomepage
+
+private const val THEME_ID = "theme-id"
+private const val THEME_NAME = "Awesome Theme"
+private const val RESOURCE_STRING = "This is a resource string"
+
+@InternalCoroutinesApi
+class ThemesViewModelTest : BaseUnitTest() {
+    @Mock lateinit var themeStore: ThemeStore
+    @Mock lateinit var resourceProvider: ResourceProvider
+    @Mock lateinit var site: SiteModel
+    private lateinit var viewModel: ThemesViewModel
+
+    val state: BottomSheetUIState get() = viewModel.bottomSheetUiState.value!!
+    val action: BottomSheetAction? get() = viewModel.bottomSheetAction.value
+
+    @Before
+    fun setUp() {
+        viewModel = ThemesViewModel(
+                themeStore,
+                resourceProvider,
+                TEST_DISPATCHER
+        )
+        startViewModel()
+        setupResourceProvider()
+    }
+
+    @Test
+    fun `given WPCom api not used, when activating theme, then don't show bottom sheet`() {
+        // Arrange
+        whenever(site.isUsingWpComRestApi).thenReturn(false)
+
+        // Act
+        viewModel.onActivateMenuItemClicked(THEME_ID)
+
+        // Assert
+        assertNull(action)
+    }
+
+    @Test
+    fun `given theme not found in installed themes, when activating theme, then search in WPCom themes`() {
+        // Arrange
+        whenever(site.isUsingWpComRestApi).thenReturn(true)
+        whenever(themeStore.getInstalledThemeByThemeId(site, THEME_ID)).thenReturn(null)
+
+        // Act
+        viewModel.onActivateMenuItemClicked(THEME_ID)
+
+        // Assert
+        verify(themeStore).getWpComThemeByThemeId(THEME_ID)
```

Comment:
We can enhance the `BottomSheetAction` to hold the value of the theme and check this value here instead.

Reply:
I am not sure I follow. No `BottomSheetAction` will be fired in this test. I am basically testing [this](https://github.com/kashandanish/WordPress-Android/blob/35483d2e6b79a95ca87cad659341ef4390d4a594/WordPress/src/main/java/org/wordpress/android/viewmodel/themes/ThemesViewModel.kt#L52) line. Are you suggesting instead of just `verify`ing this call, I should `mock` the result and proceed to the end of the method till `bottomSheetAction.value = Show`?

Reply:
We first try getting `theme` object from `getInstalledThemeByThemeId` , if not found (i.e., result is `null`) then we try getting it from `getWpComThemeByThemeId`. Translated [this](https://github.com/kashandanish/WordPress-Android/blob/35483d2e6b79a95ca87cad659341ef4390d4a594/WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserActivity.java#L366) code to Kotlin using Elvis operator.

Reply:
Ok. But the new `theme` value assigned with the call to `onActivateMenuItemClicked` [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15043/files#diff-8841b1bbcfeebd74dd8b30666e827ea0ce04bf8a2a7856c02fc730e794d55726R52) does not change any value in the `themeStore` and [is only used for setting the text](https://github.com/wordpress-mobile/WordPress-Android/pull/15043/files#diff-8841b1bbcfeebd74dd8b30666e827ea0ce04bf8a2a7856c02fc730e794d55726R59), correct?
I still not get how the the assertion works in this test 🤷

Reply:
Alright, let me take a step back and explain this from the start.
**What I am trying to test is:** If the theme is not installed (`themeStore.getInstalledThemeByThemeId` returns `null`) then the app tries to get is from `themeStore.getWpComThemeByThemeId`. That is all I wanted to test.
**No `bottomSheetAction` is fired in this test because:** `themeStrore` is a mock and I did not use `whenever(themeStore.getWpComThemeByThemeId).thenReturn(aThemeObject)`. Because I just wanted to verify if the app tries to get theme from 2nd method if is not found via 1st method.

Reply:
Thank you for the detailed explanation @kashandanish 🙇 
After our outside discussion I understand that this tests the verifies that the `getWpComThemeByThemeId` is fired once 👍